fairmq/FairMQDevice.cxx

```diff
@@ -507,7 +507,7 @@ void FairMQDevice::RunWrapper()
                 HandleMultipleChannelInput();
             }
         } else {
-            tools::RateLimiter rateLimiter(fRate);
+            static thread_local tools::RateLimiter rateLimiter(fRate);
```

Comment:
This changes semantics: The time between `maybe_sleep` calls is measured and determines the new sleep time. It would need an explicit `reset` function to work (which would need to call `clock::now`). But then, why not just construct it on the stack? It should be more efficient on the stack anyway than when it's in global memory (and a new `reset` function gets called).

Reply:
How about:
```c++
std::optional<tools::RateLimiter> rateLimiter = [&] {
  if (fRate > 0.001) {
    return std::optional<tools::RateLimiter>(std::in_place_t, fRate);
  } else {
    return std::nullopt_t;
  }
}();
// ...
if (rateLimiter) {
  rateLimiter->maybe_sleep();
}
```

Reply:
You're right. Having quick look at `maybe_sleep()` I thought the first invocation might fall through without seeping bc the timestamp from constructor would be too old. However, every subsequent invocation (within the same loop) should do the rate-limiting just fine? Perhaps a stack variable with `0` start time would work better, and `maybe_sleep()` should handle that case (to avoid adding `reset()`)

Reply:
If the rate limiter ctor doesn't record a time, the first call to `maybe_sleep` has no initial guess for the correct sleep time / back off rate. So it would need an additional state: constructed -> initialized -> rate limiting. Where currently constructed and initialized are combined in the ctor. This optimizes the rate limiter for the cases where it's actually used. It seems you noticed calls to `clock::now` where the rate limiter wasn't actually used, right?

Reply:
That's right, the `clock::now` from constructor was called when rate limiting was not used at all. I meant the initial sleep guess could be derived from the rate parameter itself, and then improved with subsequent `maybe_sleep` calls.

Reply:
But that would still optimize the rate limiter for cases where it's not used. It should be optimized for cases where it's used. (Besides that, the ctor already does a guess from the rate parameter, the first maybe_sleep call then improves on it. It's non-trivial to change & verify how this class works.)
I believe the correct fix is to not use the rate limiter when it's not used. :wink: I.e. don't construct objects you won't need.